diff --git a/SPIRV/GlslangToSpv.cpp b/SPIRV/GlslangToSpv.cpp
index 5c2f4094..118f8527 100644
--- a/SPIRV/GlslangToSpv.cpp
+++ b/SPIRV/GlslangToSpv.cpp
@@ -10830,6 +10830,14 @@ spv::Id TGlslangToSpvTraverser::getSymbolId(const glslang::TIntermSymbol* symbol
         // default to 0
         builder.addDecoration(id, spv::Decoration::Binding, 0);
     }
+    if (options.allowNonStandardDecorations) { // axslcc spec
+        if (symbol->getQualifier().hasSamplerSlot())
+            builder.addDecoration(id, spv::Decoration::SamplerSlot, symbol->getQualifier().layoutSamplerSlot);
+        else if (IsDescriptorResource(symbol->getType())) {
+            // default to 0
+            builder.addDecoration(id, spv::Decoration::SamplerSlot, 0);
+        }
+    }
     if (symbol->getQualifier().hasAttachment())
         builder.addDecoration(id, spv::Decoration::InputAttachmentIndex, symbol->getQualifier().layoutAttachment);
     if (glslangIntermediate->getXfbMode()) {
diff --git a/SPIRV/GlslangToSpv.h b/SPIRV/GlslangToSpv.h
index 9fb4f3ff..e12dcc26 100644
--- a/SPIRV/GlslangToSpv.h
+++ b/SPIRV/GlslangToSpv.h
@@ -55,6 +55,7 @@ struct SpvOptions {
     bool emitNonSemanticShaderDebugSource{ false };
     bool compileOnly{false};
     bool optimizerAllowExpandedIDBound{false};
+    bool allowNonStandardDecorations{false}; // axslcc spec
 };
 
 GLSLANG_EXPORT void GetSpirvVersion(std::string&);
diff --git a/SPIRV/doc.cpp b/SPIRV/doc.cpp
index dcc9b4e1..724bc141 100644
--- a/SPIRV/doc.cpp
+++ b/SPIRV/doc.cpp
@@ -350,6 +350,7 @@ const char* DecorationString(int decoration)
     case (int)Decoration::HitObjectShaderRecordBufferEXT:  return "DecorationHitObjectShaderRecordBufferEXT";
 
     case (int)Decoration::SaturatedToLargestFloat8NormalConversionEXT: return "DecorationSaturatedToLargestFloat8NormalConversionEXT";
+    case 7000: return "SamplerSlot"; // axslcc spec
     }
 }
 
diff --git a/SPIRV/spirv.hpp11 b/SPIRV/spirv.hpp11
index 6f3021e9..57dfe1a1 100644
--- a/SPIRV/spirv.hpp11
+++ b/SPIRV/spirv.hpp11
@@ -206,7 +206,7 @@ enum class ExecutionMode : unsigned {
     SampleInterlockOrderedEXT = 5368,
     SampleInterlockUnorderedEXT = 5369,
     ShadingRateInterlockOrderedEXT = 5370,
-    ShadingRateInterlockUnorderedEXT = 5371,
+    ShadingRateInterlockUnorderedEXT = 5371,
     Shader64BitIndexingEXT = 5427,
     SharedLocalMemorySizeINTEL = 5618,
     RoundingModeRTPINTEL = 5620,
@@ -653,6 +653,7 @@ enum class Decoration : unsigned {
     ImplementInRegisterMapINTEL = 6191,
     CacheControlLoadINTEL = 6442,
     CacheControlStoreINTEL = 6443,
+    SamplerSlot = 7000, // axslcc spec
     Max = 0x7fffffff,
 };
 
@@ -1199,7 +1200,7 @@ enum class Capability : unsigned {
     RayTracingDisplacementMicromapNV = 5409,
     RawAccessChainsNV = 5414,
     RayTracingSpheresGeometryNV = 5418,
-    RayTracingLinearSweptSpheresGeometryNV = 5419,
+    RayTracingLinearSweptSpheresGeometryNV = 5419,
     Shader64BitIndexingEXT = 5426,
     CooperativeMatrixReductionsNV = 5430,
     CooperativeMatrixConversionsNV = 5431,
@@ -3426,7 +3427,7 @@ inline const char* ExecutionModeToString(ExecutionMode value) {
     case ExecutionMode::SampleInterlockOrderedEXT: return "SampleInterlockOrderedEXT";
     case ExecutionMode::SampleInterlockUnorderedEXT: return "SampleInterlockUnorderedEXT";
     case ExecutionMode::ShadingRateInterlockOrderedEXT: return "ShadingRateInterlockOrderedEXT";
-    case ExecutionMode::ShadingRateInterlockUnorderedEXT: return "ShadingRateInterlockUnorderedEXT";
+    case ExecutionMode::ShadingRateInterlockUnorderedEXT: return "ShadingRateInterlockUnorderedEXT";
     case ExecutionMode::Shader64BitIndexingEXT: return "Shader64BitIndexingEXT";
     case ExecutionMode::SharedLocalMemorySizeINTEL: return "SharedLocalMemorySizeINTEL";
     case ExecutionMode::RoundingModeRTPINTEL: return "RoundingModeRTPINTEL";
@@ -4157,7 +4158,7 @@ inline const char* CapabilityToString(Capability value) {
     case Capability::RayTracingDisplacementMicromapNV: return "RayTracingDisplacementMicromapNV";
     case Capability::RawAccessChainsNV: return "RawAccessChainsNV";
     case Capability::RayTracingSpheresGeometryNV: return "RayTracingSpheresGeometryNV";
-    case Capability::RayTracingLinearSweptSpheresGeometryNV: return "RayTracingLinearSweptSpheresGeometryNV";
+    case Capability::RayTracingLinearSweptSpheresGeometryNV: return "RayTracingLinearSweptSpheresGeometryNV";
     case Capability::Shader64BitIndexingEXT: return "Shader64BitIndexingEXT";
     case Capability::CooperativeMatrixReductionsNV: return "CooperativeMatrixReductionsNV";
     case Capability::CooperativeMatrixConversionsNV: return "CooperativeMatrixConversionsNV";
diff --git a/glslang/Include/Types.h b/glslang/Include/Types.h
index d3124ac8..5580ef25 100644
--- a/glslang/Include/Types.h
+++ b/glslang/Include/Types.h
@@ -929,6 +929,9 @@ public:
                  unsigned int layoutBinding              : 16;
     static const unsigned int layoutBindingEnd      =  0xFFFF;
 
+                     unsigned int layoutSamplerSlot      : 16;
+    static const unsigned int layoutSamplerSlotEnd   = 0xFFFF;
+
                  unsigned int layoutIndex                :  8;
     static const unsigned int layoutIndexEnd      =      0xFF;
 
@@ -994,6 +997,7 @@ public:
 
         layoutSet = layoutSetEnd;
         layoutBinding = layoutBindingEnd;
+        layoutSamplerSlot = layoutSamplerSlotEnd;
         layoutAttachment = layoutAttachmentEnd;
     }
 
@@ -1027,6 +1031,10 @@ public:
     {
         return layoutBinding != layoutBindingEnd;
     }
+    bool hasSamplerSlot() const
+    {
+        return layoutSamplerSlot != layoutSamplerSlotEnd;
+    }
     bool hasOffset() const
     {
         return layoutOffset != layoutNotSet;
diff --git a/glslang/Include/glslang_c_interface.h b/glslang/Include/glslang_c_interface.h
index 4eec55b5..caa880fd 100644
--- a/glslang/Include/glslang_c_interface.h
+++ b/glslang/Include/glslang_c_interface.h
@@ -239,6 +239,7 @@ typedef struct glslang_spv_options_s {
     bool emit_nonsemantic_shader_debug_source;
     bool compile_only;
     bool optimize_allow_expanded_id_bound;
+    bool allow_non_standard_decorations; // axslcc spec
 } glslang_spv_options_t;
 
 #ifdef __cplusplus
diff --git a/glslang/MachineIndependent/ParseHelper.cpp b/glslang/MachineIndependent/ParseHelper.cpp
index 8f1a3817..98160a6e 100644
--- a/glslang/MachineIndependent/ParseHelper.cpp
+++ b/glslang/MachineIndependent/ParseHelper.cpp
@@ -7070,6 +7070,10 @@ void TParseContext::setLayoutQualifier(const TSourceLoc& loc, TPublicType& publi
             error(loc, "needs a literal integer", "binding", "");
         return;
     }
+    else if (id == "sampler_slot") { // axslcc spec
+        publicType.qualifier.layoutSamplerSlot = value;
+        return;
+    }
     if (id == "constant_id") {
         requireSpv(loc, "constant_id");
         if (value >= (int)TQualifier::layoutSpecConstantIdEnd) {
@@ -7408,6 +7412,8 @@ void TParseContext::mergeObjectLayoutQualifiers(TQualifier& dst, const TQualifie
             dst.layoutSet = src.layoutSet;
         if (src.layoutBinding != TQualifier::layoutBindingEnd)
             dst.layoutBinding = src.layoutBinding;
+        if (src.layoutSamplerSlot != TQualifier::layoutSamplerSlotEnd)
+            dst.layoutSamplerSlot = src.layoutSamplerSlot;
 
         if (src.hasSpecConstantId())
             dst.layoutSpecConstantId = src.layoutSpecConstantId;
