diff --git a/3rdparty/spirv-cross/spirv_glsl.cpp b/3rdparty/spirv-cross/spirv_glsl.cpp
index 81fdae0..f7006b7 100644
--- a/3rdparty/spirv-cross/spirv_glsl.cpp
+++ b/3rdparty/spirv-cross/spirv_glsl.cpp
@@ -2373,7 +2373,10 @@ void CompilerGLSL::emit_buffer_block(const SPIRVariable &var)
 		emit_buffer_block_flattened(var);
 	else if (is_legacy() || (!options.es && options.version == 130) ||
 	         (ubo_block && options.emit_uniform_buffer_as_plain_uniforms))
-		emit_buffer_block_legacy(var);
+		if (options.inline_ubo_members) // axslcc spec
+			emit_buffer_block_inlined(var);
+		else
+			emit_buffer_block_legacy(var);
 	else
 		emit_buffer_block_native(var);
 }
@@ -10826,6 +10829,8 @@ string CompilerGLSL::access_chain_internal(uint32_t base, const uint32_t *indice
 					string qual_mbr_name = get_member_qualified_name(type_id, index);
 					if (!qual_mbr_name.empty())
 						expr = qual_mbr_name;
+					else if (options.inline_ubo_members) // axslcc spec
+						expr = to_member_name(*type, index);
 					else if (flatten_member_reference)
 						expr += join("_", to_member_name(*type, index));
 					else
@@ -20219,3 +20224,37 @@ bool CompilerGLSL::has_legacy_nocontract(uint32_t result_type, uint32_t id) cons
 	                      FPFastMathModeAllowReassocMask;
 	return (get_fp_fast_math_flags_for_op(result_type, id) & fp_flags) != fp_flags;
 }
+
+// --- axslcc spec
+void CompilerGLSL::emit_buffer_block_inlined(const SPIRVariable& var)
+{
+	auto &type = get<SPIRType>(var.basetype);
+	bool ssbo = var.storage == StorageClassStorageBuffer ||
+	            ir.meta[type.self].decoration.decoration_flags.get(DecorationBufferBlock);
+	if (ssbo)
+		SPIRV_CROSS_THROW("SSBOs not supported in legacy targets.");
+
+	// We're emitting the push constant block as a regular struct, so disable the block qualifier temporarily.
+	// Otherwise, we will end up emitting layout() qualifiers on naked structs which is not allowed.
+	auto &block_flags = ir.meta[type.self].decoration.decoration_flags;
+	bool block_flag = block_flags.get(DecorationBlock);
+	block_flags.clear(DecorationBlock);
+
+	add_resource_name(type.self);
+	auto name = type_to_glsl(type);
+	type.member_name_cache.clear();
+	uint32_t i = 0;
+	for (auto &member : type.member_types)
+	{
+		add_member_name(type, i);
+		statement_inner("uniform ");
+		emit_struct_member(type, member, i);
+		++i;
+	}
+	statement("");
+
+	if (block_flag)
+		block_flags.set(DecorationBlock);
+}
+
+// === axslcc spec
diff --git a/3rdparty/spirv-cross/spirv_glsl.hpp b/3rdparty/spirv-cross/spirv_glsl.hpp
index 78bff2d..a5914b0 100644
--- a/3rdparty/spirv-cross/spirv_glsl.hpp
+++ b/3rdparty/spirv-cross/spirv_glsl.hpp
@@ -156,6 +156,18 @@ public:
 		// This workaround may cause significant performance degeneration on some Android devices.
 		bool enable_row_major_load_workaround = true;
 
+		// axslcc spec
+		// Converts uniform block members into individual global uniform declarations.
+		// Example:
+		//    uniform fs_ub {
+		//       float a;
+		//       float b;
+		//    }
+		// becomes:
+		//    uniform float a;
+		//    uniform float b;
+		bool inline_ubo_members = false;
+
 		// If non-zero, controls layout(num_views = N) in; in GL_OVR_multiview2.
 		uint32_t ovr_multiview_view_count = 0;
 
@@ -1085,6 +1097,11 @@ protected:
 	uint32_t get_fp_fast_math_flags_for_op(uint32_t result_type, uint32_t id) const;
 	bool has_legacy_nocontract(uint32_t result_type, uint32_t id) const;
 
+	// --- axslcc spec: inline ubo members support
+public:
+	void emit_buffer_block_inlined(const SPIRVariable &var);
+	// === axslcc spec
+
 private:
 	void init();
 
